src/Database/Esqueleto/PostgreSQL.hs

```diff
@@ -477,3 +480,49 @@ forUpdateOf lockableEntities onLockedBehavior =
 forShareOf :: LockableEntity a => a -> OnLockedBehavior -> SqlQuery ()
 forShareOf lockableEntities onLockedBehavior =
   putLocking $ PostgresLockingClauses [PostgresLockingKind PostgresForShare (Just $ LockingOfClause lockableEntities) onLockedBehavior]
+
+-- | `UPDATE .. RETURNING ..` SQL extension supported by Postgres.
+--
+-- The instances of 'InferReturning' say what can be returned; currently includes
+-- whole entities, 'PersistField's, SQL expressions, tuples (possibly nested).
+--
+-- Usage example:
+--
+-- @
+--   tuples <- updateReturning $ \p -> do
+--     set p [ PersonAge =. val (Just 0) ]
+--     where_ (isNothing $ p ^. PersonAge)
+--     return (val True, p ^. PersonName, (p ^. PersonFavNum) *. val 100)
+--     -- return p -- also works, returning (Entity Person)
+-- @
+--
+-- @since 3.5.11.3
+updateReturning :: (MonadIO m, From from, InferReturning ex ret, SqlBackendCanWrite backend)
+                => (from -> SqlQuery ex)
+                -> R.ReaderT backend m [ret]
+updateReturning block = do
+  conn <- R.ask
+  conduit <- rawSelectSource UPDATE (tellReturning ReturningStar >> from block)
+  liftIO . withAcquire conduit $ flip R.runReaderT conn . runSource
```

Comment:
So this is the *use* of the new functionality - neat.

This is where we say "Do the `RETURNING` thing":

```haskell
rawSelectSource UPDATE (tellReturning ReturningStar >> from block)
```

So we modify the `SideData` for the query, then it gets rendered and turned into a `conduit`, which we run.

I think we can *avoid* modifying `SideData` and still mostly keep this interface.

Consider [`renderQueryToText`](https://github.com/bitemyapp/esqueleto/blob/30a5e80736391e2aa45094f681d4bd329aa16707/src/Database/Esqueleto/Internal/Internal.hs?plain=1#L3005-L3024). This takes a `SqlQuery a` and renders it out.

If the `SqlQuery a` has a `ReturningStar` in the `SideData`, then we graft on the appropriate text to the query. However, we can just *do that* directly without modifying the internals and introducing potential unsafety.

If we look at the code for [`rawSelectSource`](https://github.com/bitemyapp/esqueleto/blob/30a5e80736391e2aa45094f681d4bd329aa16707/src/Database/Esqueleto/Internal/Internal.hs?plain=1#L2717-L2745), we can see that it's just doing `toRawSql` on the query and then `rawQueryRes`. The rest is handling the `Conduit` (which imo isn't super useful since the query must be loaded into memory anyway).

So, if we inline this and use `rawQuery` instead (which just does the list directly), we can do:

```haskell
updateReturning block = do
    conn <- R.ask
    let 
        query' = toRawSql UPDATE (conn, initialIdentState) block
        ((ret, _sd), _finalIdent) = flip S.runState -- ... (from toRawSql) 
        returningClause =  ("RETURNING ", []) <> sqlSelectCols info ret
        (queryWithReturning, vals) = 
             query' <> ("\n", []) <> returningClause
    rawQuery queryWithReturning vals
```

Now, it *is* unsatisfying that we have to run the query building twice with this - once in `toRawSql` and once to get the `ret` to make the `returningClause` with.

If `toRawSql` is refactored into a variant that *also* returns the `ret`, then I think we're able to avoid that - then this can be:

```haskell
    conn <- R.ask
    let 
        (ret, query') = toRawSqlWith UPDATE (conn, initialIdentState) block
        returningClause =  ("\nRETURNING ", []) <> sqlSelectCols info ret
        (queryWithReturning, vals) = 
             query' <> returningClause
    rawQuery queryWithReturning vals
```

Then I think there's probably virtue in writing `makeReturning :: (InferReturning a r) => a -> (TLB.Builder, [PersistValue])` which can then be reused in `deleteReturning`.